### PR TITLE
Fix inconsistent multireturn values for stored C functions

### PR DESCRIPTION
**Pull request by Buristan**
The first commit introduces the `c_func_iproto_multireturn` option (when set to new, the results of the stored C function aren't wrapped in an additional msgpack array, see below). By default, it is set to 'old', to avoid conflicts.  Also, this makes the first patch backportable to 2.11 as is (except conflict with compatibility options in compat.lua, see #9144). The second commit switches the behaviour to 'new' and fixes some diff tests.

As described in the commit message to the first patch, the SQL mechanism depends on the `port_dump_msgpack()` too, so just using the introduced flag in `port_c_dump_msgpack()` isn't good. As an alternative option, we can introduce 2 new fields (say `encode_call_results16`, `encode_call_results`) in `port_vtab` to call them and use the flag only there. But I'm not sure that adding 2 new fields to the structure only for this is good either.

As far as there are only two possible fields in the changelog header entry (`feature/*` or `bugfix/*`), I prefer the second one instead of `compatibility/*`.

I provide the doc-request below for your convenience in reading:

---

Title: Document `c_func_iproto_multireturn` compat option

Please create a documentation page for the new compat option:
https://tarantool.io/compat/c_func_iproto_multireturn

In the new behaviour, the multiresults returned by a stored C function
via iproto aren't wrapped in the additional msgpack array (old).

```
tarantool> compat.c_func_iproto_multireturn = 'old'
---
...

tarantool> net_box.connect(box.cfg.listen):call('myclib.cfunc')
---
- [true, -1]
...

tarantool> compat.c_func_iproto_multireturn = 'new'
---
...

tarantool> net_box.connect(box.cfg.listen):call('myclib.cfunc')
---
- true
- -1
...

```

The new behaviour is consistent with the local call of the function
via `box.func`:

```
tarantool> box.func['myclib.cfunc']:call()
---
- true
- -1
...

```

Assume you have a stored C function that returns values like the
following:

```c
char *position = mp_encode_bool(buffer, true);
box_return_mp(ctx, buffer, position);
/* ... */
position = mp_encode_int(buffer, -1);
box_return_mp(ctx, buffer, position);
```

If you want to preserve the format of the returned array for your C
functions, when the `c_func_iproto_multireturn` option is set to "new",
you should add the additional wrapping, like the following:

```c
char *position = mp_encode_array(buffer_with_results, n_results);
position = mp_encode_bool(position, true);
/* ... */
position = mp_encode_int(position, -1);
box_return_mp(ctx, buffer_with_results, position);
```

The amount of `box_return_mp()` calls indicates the number of values to
be returned.

Also, you should update its usage via `box.func` if there is any.

### CODE REVIEW

**Review by ochaplashkin — APPROVED** [review]
LGTM :rocket:

File: src/box/iproto.cc [hunk 1]
```diff
@@ -2322,7 +2322,8 @@ tx_process_call(struct cmsg *m)
 	out = msg->connection->tx.p_obuf;
 	iproto_prepare_select(out, &svp);
 
-	if (msg->header.type == IPROTO_CALL_16)
+	if (msg->header.type == IPROTO_CALL_16 ||
+	    port_c_unwrap_multireturn(&port))
```
Comment (Totktonada) [inline src/box/iproto.cc:2326]: It seems, we can get rid of iproto.cc changes by adjusting `port_sql_dump_msgpack` function. It is a bit messy, but this mess will be at the lower level.

I see no test fails regarding box.session.push/box_session_push if I change port_c_dump_msgpack directly. Please, re-check this point.

Comment (Buristan) [reply src/box/iproto.cc]: Yes. Updated the branch. Also, I renamed functions, as we discussed offline, and added a note about the amount of `box_return_mp()` in the doc request.

`box.session.push()` uses the Lua port, so these changes are unrelated.

File: changelogs/unreleased/gh-4799-consistent-c-functions-calls.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+## bugfix/box
+
+* **[Breaking change]** Added a `c_func_iproto_multireturn` option to
+  tarantool.compat. The new behaviour drops an additional array that wraps
+  multiple results returned via iproto (gh-4799).
```
Comment (xuniq) [inline changelogs/unreleased/gh-4799-consistent-c-functions-calls.md:5]:
```suggestion
* **[Breaking change]** Added a `c_func_iproto_multireturn` option to the
  `compat` module. The new behavior drops an additional array that wraps
  multiple results returned via iproto (gh-4799).
```

**Review by xuniq — APPROVED** [review]

**Review by Totktonada — APPROVED** [review]
I'm OK with the implementation if the approach with changing the default behavior for existing code (including built *.so/*.dylib) is the best that we can do.

I would like some more gentle upgrade path, but I don't see any variant that would solve the main problem: consistency between iproto call and box.func call (plus consistency with iproto/box.func calls of a Lua code).

If anyone has ideas here, I would appreciate them.

**Review by Gerold103 — APPROVED** [review]
Thanks for the patch! All looks good.

**Pull request merged**